### ISSUE TITLE
Using Error with RWS code generation causes logs and exception to be lost

### ISSUE THREAD

**Issue by colethecoder** [louthy/language-ext#634]
Within a .Net Core Console App I created a simple RWS monad using the CodeGen package and the following code:

```cs
    public class Person
    {
        public readonly string Forename;
        public readonly string Surname;

        public Person(string forename, string surname)
        {
            Forename = forename;
            Surname = surname;
        }
    }

    public class IO
    {
        public Person ReadFromDB() => new Person("Spider", "Man");
    }

    [RWS(WriterMonoid: typeof(MSeq<string>),
         Env: typeof(IO),
         State: typeof(Person),
         Constructor: "Pure",
         Fail: "Error")]
    public partial struct Subsys<T>
    {
    }
```

and then try to use it with:

```cs
        static void Main(string[] args)
        {
            var makeName =
                         from _1 in Subsys.tell(Seq1("Started"))
                         from x in Subsys.ReadFromDB()
                         from _2 in Subsys.tell(Seq1("Got from DB"))
                         from g in x.Forename == "Tom"
                                    ? Subsys.Pure(unit)
                                    : Subsys.Error<Unit>("Not Me!!", new Exception("NOT ME"))
                         from y in Subsys.put(x)
                         from _3 in Subsys.tell(Seq1("Updated state"))
                         from a in Subsys.Forename
                         from b in Subsys.Surname
                         from _4 in Subsys.tell(Seq1("Got name parts"))
                         select $"{a} {b}";

            var result = makeName.Run(new IO(), new Person("Foo", "Bar"));

            Console.WriteLine("Logs:");
            result.Output.Iter(x => Console.WriteLine(x));

            Console.WriteLine("Name:");
            result.Value.Match(
                Some: x => Console.WriteLine(x),
                None: () => Console.WriteLine("NOTHING"),
                Fail: x => Console.WriteLine(x.Message));

            Console.ReadLine();
        }
```

My expectation is that the output would be:

```
Logs:
Started
Got from DB
Name:
NOT ME
```

However the actual output is:

```
Logs:
Name:
Object reference not set to an instance of an object.
```

I believe the issue is caused by the Bind function in MRWS.cs not being able to check whether the state IsFaulted before it attempts to calculate `next` which then causes a null reference exception which is caught in `RWS.Run` and wipes out the writer state as well as overwriting the exception state.

I'll have a look at fixing this but it doesn't seem trivial at first glance.

**Comment by louthy** [comment]
@colethecoder I'll take a look

**Comment by louthy** [comment]
Tom, this is fixed in `v3.3.25-beta`.  I've marked it as a beta because whilst I was in there I did some optimisations of the `RWS` monad and also optimised the `Seq.Concat` method which is used a lot for logging if you use an `MSeq<A>` monoid as the `MonoidW` generic parameter.  

It can now spot if two strict sequences are being concatenated (which is usually the case when using `RWS` or `Writer`) and try and grow into the pre-allocated array backing the strict-Seq, so there'll be less churn in terms of memory allocations.

Internally as well as `SeqLazy<A>`, `SeqStrict<A>`, and `SeqEmpty<A>`, there's now a `SeqConcat<A>` which waits for as long as possible before flattening any sequences.  This is useful when concatenating lazy sequences.  However, because there's now 4^2 possible combinations of internal sequences (when concatenating) I think it's best to release this as a beta until I've done some deeper testing.

This should have some tangible performance benefits for the `Validation` monad also.

If you find any issues, please let me know.
